### DISCUSSION TITLE
Brunch on MacBooks - Please read :)

### DISCUSSION THREAD

**Discussion by ZProLegend007** [sebanc/brunch#1926] [General]
I have a MacBook (2018 Pro T2 Touchbar) with a pentaboot setup. It runs rEFInd to boot into MacOS, Windows 11, ZorinOS, Kali Linux and Grub. Grub boots Brunch.

I have done some extensive troubleshooting and research to get brunch working as much as I can on my MacBook. I have made an issue detailing my documentation on fixes and issues that need to be solved. I mentioned sebanc but am yet to get a response, understood as he is likely very busy. The T2 kernel was updated in brunch-unstable which may be some indication he has seen the issue. If anyone has a MacBook and wants to get things running or you are willing to give some support so this can hopefully gain some momentum and move up sebanc's to-do list then any comments, ideas, thoughts, encouragement or reactions are appreciated. I'll link the issue below, please give it a look, I'm in high school and this sort of stuff is a lot of work :)

Once we fix the major issues then we can look to documentation and instructions in the wiki or readme.md if needed to include guides for people using MacBooks.

See documentation and issue here: https://github.com/sebanc/brunch/issues/1914

## A repo designed for brunch on MacBooks is linked here!! [brunch-mac](https://github.com/ZProLegend007/brunch-mac). I will be working on this non-stop trying to solve as many issues as possible. Check it out!

**Comment by DamianNiolet** [comment]
I want to express my gratitude to @ZProLegend007!  Great work!  If the audio is the only lingering issue, even if on an older release, and it's mostly stable, that's awesome!  Also want to express my desire for additional support from the community.  Let's make this happen!

I am not a coder, but I know how to follow guides and I'm more than willing to assist with testing.  I have multiple Macbook Pros I can test on - 2012 (with Nvidia), 2014 (without Nvidia), 2015 (with Nvidia), 2019 (with AMD GPU), 2020 (without dGPU).  I know this conversation focuses on T2 Macs, but might be useful to have a baseline without T2 chips.

**Reply by ZProLegend007** [reply]
Hey! Thank you! I have been away for the past week and just got back. That would be great and seeing some support is and would be awesome! 

I believe most the older non-T2 Mac's work with a normal kernel as they contain less proprietary components, and the T2 MacBooks have their own kernel in brunch. If you could test some solutions on your T2 Mac's that would be greatly appreciated. I'll have a look soon and get back to this.

**Reply by ZProLegend007** [reply]
Could you test the macbook-t2 kernel on both your 2019 and 2020 Macbooks? and then please let me know what works and what doesn't on those in the #1914 issue, that would be awesome and we could go from there. Thanks again.

**Reply by DamianNiolet** [reply]
@ZProLegend007, happy to!  However, I just followed the dual-booting install instructions exactly, but rEFInd isn't finding Brunch.  I have Ubuntu on here too, and I updated grub according to the instructions, but the grub bootloader goes straight to Ubuntu, never even showing me the options.  But, when I use grub customizer, I see the Brunch menuentries...I even tried putting the same menuentries into rEFInd to no avail.  So, I'm stuck getting the bootloader to work before I can test the install.  Sorry, this is my first time installing brunch on antything.

**Reply by ZProLegend007** [reply]
All good, rEFInd doesn't have the capabilities to boot brunch (I believe). Did you use the dual-boot instructions in the wiki? It's not super helpful IMO but if it worked for you that's good. 

In grub customiser make sure that grub is in menu mode not hidden mode. Also make sure the timeout is at least 1 second. 

I assume you are using the latest r116 release? Make sure to add the correct kernel parameters while changing the brunch kernel and settings.

**Reply by DamianNiolet** [reply]
Yeah, using r116.  And yes, followed the "Install with Linux" guide.  I have to change the kernels by booting into the brunch settings, correct?

Still not able to get the grub menu to show, even after using grub customizer, I assume because it's not actually updating the rEFInd files.  The menuentries I'm making for rEFInd.conf aren't doing any good either.  I have Windows on here too.  And I know grub was updated after installing Brunch because the entries are there in grub customizer and it killed my Windows bootloader.  I had to use gdisk to convert the partition table from Hybrid to Protective to get Windows booting again.  So, rEFInd will only show Windows, Ubuntu, and MacOS, but Brunch is definitely in place.  Any suggestions?

**Reply by DamianNiolet** [reply]
OK, so...I was able to get into Brunch settings by manually editing the grub.cfg file on my Ubuntu partition (changing the default to "Brunch settings").  I then followed your steps as far as the settings, but ran into issues right away:

1) trackpad still not working.  Mouse with dongle overcomes this though.
2) crosh did not recognize any of the commands to get wifi working.  I already had the wifi firmware on EFI, from installing Ubuntu before, but crosh doesn't even have "rm" or "mv" or even "cd" in its commands list.  Ethernet dongle overcomes this too and I was able to update, but no changes to the issues.
3) No touchbar
4) No audio, but that was to be expected

Of course, issue zero is getting my grub to work, so I don't have to keep editing grub.cfg in MacOS every time I want to go between Ubuntu and ChromeOS.

All of this is on the 2020 13" MBpro with 4 ports, 2.3 i7, 32GB  RAM, BTW

**Reply by ZProLegend007** [reply]
Ah, I see. I was thinking of the Dual-Boot on Mac instructions which look like a pain to use. Yes that's good. Yep brunch kernel switches are done from the brunch settings. Nice job with the partition table fix, I once bricked my entire MacBook trying to fix that haha. rEFInd can't boot brunch, and it can't detect it. You need to use rEFInd to boot grub then grub to boot brunch. My rEFInd setup boots straight into linux without grub so when I installed grub I removed all the other linux OSs from the boot menu and just had brunch and brunch settings.

Try not to fix anything yet. With the Wi-Fi problem, that all good now, the main point in testing right now is to see if there are things that don't work on other MacBooks or things that do. Is there not Wi-Fi firmware present already in the r116? Also a good thing to keep in mind is that the instructions I have for fixes are designed for r103 and r107. The reason those commands aren't working is that you aren't in the shell. "_Once you are in the crosh terminal type 'shell' to enter the linux shell then 'sudo su'._" Although in newer releases "su" has been removed so you have to run commands as sudo.

My internal testing checklist is:
- Make sure the same things work/don't work consistently across all T2 MacBooks.
- Show said consistencies or inconsistencies in a comment on #1914 issue.
- Make sure the fixes (on the older releases) all work consistently on all T2 MacBooks.
- Show said consistencies or inconsistencies in another comment on #1914 issue.
- Make sure all the fixes (that hopefully @sebanc will apply) work across all T2 MacBooks.
- Close #1914 and related issues.
- Enjoy a wonderful brunch experience :)



Could you let me know if Wi-Fi can scan and detect networks and also check if you can see your battery because that is an issue for me in the latest releases, and also Suspend/Sleep. Or even better just run through the "latest release" checklist in #1914 and make sure everything is the same.

Can you take a screenshot of your grub settings in grub-customizer and I'll have a look, thanks :)

**Reply by DamianNiolet** [reply]
@ZProLegend007, I don't mean to derail this thread, but here are the screenshots from my boot situation.  I need to focus on working around this issue because I can't keep manually updating grub.cfg, updating grub, and then gdisking my partition table.  That'll get messy real quickly.  But rather than just get the grub menu to appear as it should, I've been working on adding an icon to rEFInd.  I'm so close to getting that to work!  I'll walk through the progress:

Here's a screenshot of grub customizer: ![Screenshot from 2023-10-09 18-54-05](https://github.com/sebanc/brunch/assets/146675625/480935ac-f2cc-4889-a04d-9f320c281f8b)  As you can see "show menu" is checked and "Grub_Timeout_Style" is unchecked.  And it finds the other OSs - brunch and Windows: ![image](https://github.com/sebanc/brunch/assets/146675625/104d5fc1-c3ff-483d-8581-5446d166d51b)  But in order to actually get brunch to boot I have to manually change the grub.cfg in the Ubuntu partition's /boot/grub folder.  I've gotten to the point where I have grubU.cfg, grubC.cfg, and grubS.cfg and I just remove the letter, depending on which one I need to boot into, within MacOS.

So, like I said, I didn't want to spend a lot of time trying to get the menu to appear, when the real goal for me is a separate icon for ChromeOS in rEFInd.  I did a little digging and discovered the image has an efi parition at loop14p12.  I can mount it by opening the image in Ubuntu, then using terminal, making the dir and then "$ sudo mount /dev/loop14p12 ~/tmpmount2"  I noticed that Ubuntu uses grubx64.efi in the nvme0np1/EFI/ubuntu folder, but brunch/ChromeOS uses loop14p12/efi/boot/bootx64.efi in it's own partition within the image.  So, I figured, why not copy everything within /efi and put it in /EFI?

![image](https://github.com/sebanc/brunch/assets/146675625/212d22d7-15ce-45e5-a144-f257f5738ec8)  A fourth icon now shows up in rEFInd and if you don't change anything, it even goes to the grub menu with only "ChromeOS" or "Settings;" however, neither of them actually go anywhere, typically popping errors or crashing.  At this point (not having a firm grasp on bootloaders), I'm trying to replicate the boot process into Ubuntu, since I know changing the default value in its grub.cfg file will get me to ChromeOS. But if you (or someone following along...and likely laughing) know of a more sure way to get this to work, please let me know.

Update: In the last hour I've gotten even farther, basically starting from scratch with copying efi from the image onto EFI on the drive, then fixing each error one-by-one.  The first said, "can't find disk,7" and load kernel first, so I changed all instances of "$disk" to "loop" because that's how it is in the Ubuntu grub.cfg and then copied these lines, since they aren't in the image's grub.cfg:

	rmmod tpm
	img_path=/chromeos.img
	img_uuid=b2cd24d3-dea3-4ce8-8bb4-e0a99d1474fc
	search --no-floppy --set=root --file $img_path
	loopback loop $img_path
	source (loop,12)/efi/boot/settings.cfg

But the error said it didn't recognize the "rmmod" command.  So, I just removed that top line and tried again...and Voila!  The brunch framework loads!  However, it crashes...with no error.  But clearly it can't find ChromeOS and I'm at a lose as to what to add or change from here.

**Reply by ZProLegend007** [reply]
Wow, that is very detailed well done 👍. I believe there is a simple solution. If you go to grub-customizer and then advanced. Make sure you check the `GRUB_TIMEOUT_STYLE` setting and change its value from `hidden` to `menu`. Don't worry about derailing the discussion, I'm happy to help 😁.

I'm also very impressed you got Brunch to boot (even partially) from rEFInd, that's very impressive. The reason it doesn't boot fully is (I think) because brunch uses chainloaders, which rEFInd doesn't support. There might be a setting in rEFInd for it but I doubt it.

**Reply by DamianNiolet** [reply]
I should have mentioned, I did try the “menu” thing and it still didn’t work.

Where there’s a will, there’s a way.  ;)

**Reply by ZProLegend007** [reply]
Try mimicking the relevant settings here: 

![image](https://github.com/sebanc/brunch/assets/97338659/16396b7b-e24c-4f98-b4c6-8bbf1f7b01cd)

I do believe I turned os-prober on as something broke without it. Grub timeout can stay at 10 seconds if that's what you want, I have it set to 1 so it automatically boots into brunch but I can still switch to brunch settings if needed.

**Reply by DamianNiolet** [reply]
@ZProLegend007, I reached out to Rod Smith and here's what he had to say about a standalone rEFInd entry:

"rEFInd should be able to boot any modern Linux kernel directly, *IF* that kernel includes EFI stub loader support. Placing the kernel in one of the auto-scanned locations (such as the root (/) or /boot of a filesystem that the EFI can read, either natively or via an EFI filesystem driver) should get it recognized, provided the kernel has a name that rEFInd recognizes (such as vmlinuz*)."

**Reply by ZProLegend007** [reply]
That's good to know, let me know if you manage to get it to work.

**Reply by DamianNiolet** [reply]
@ZProLegend007, I did it!

I'll write out a more complete guide, but in a nutshell, you copy everything from loop##p12 (the loop## changes based on how many other apps you have installed in linux, I discovered, (and this is ChromeOS.img's efi partition) and everything from Root-C (loop##p7), into a new /boot folder on the ChromeOS partition.  Then create a grub.cfg in a folder on rEFInd (mine is "chromeos"), making the file similar to the one pointing to your linux installation (mine being "ubuntu"), but change the UUID and grub.cfg path to point to ChromeOS/boot, then adjust the grub.cfg file you copied over to ChromeOS partition/boot, by removing the existing entries and adding in the entries created during installation, which are in an auto-generated "chromes.img.grub.txt" file on the ChromeOS partition.  Lastly, remove the line, "rmmod tmp" from both entries.  It would still work with those lines, but removes the error that pops up when booting.

Happy booting!

**Comment by DamianNiolet** [comment]
@ZProLegend007, been at it still, but messed stuff up so much in the process, I had to reinstall Ubuntu to clean it up and try again with brunch.  I think there may have been an issue with my last install of Ubuntu, because are going much better now.  What was the issue exactly, IDK, but I think it had something to do with permissions.

At any rate, I did figure out the issue with my grub menu...sadly, the problem was rEFInd.  The menu was there the whole time, but hidden behind rEFInd's boot screen - the one with the blue bar and two lines of code.  I realized this after setting grub to 10 sec timeout and that boot screen stayed up for exactly 10 secs.  Then I tried the down arrow (knowing brunch was just one entry down) and viola...Still trying to figure out what I need to change in rEFInd to make the grub menu appear in front of that screen, but at least I don't have to mess with .cfg files anymore.

I'm surprised at the lack of interest in this fork.  t2 macs are going to be around for a while still and I'm sure there are a decent number of people who would like to have ChromeOS on theirs, even if they aren't coders or github users.  I'll follow along and help where I can.  I have actual brunch updates to provide in the other fork, having gone back to r117.

**Reply by ZProLegend007** [reply]
Coolio, I can help you out with rEFInd there. Make sure you have set grub to graphical boot in refind.conf in your EFI partition.
